Movement

ChristopherKallstrom opened the pull request.

ChristopherKallstrom committed:
Add acceleration for horizontal player movement and sprint functionality

ChristopherKallstrom committed:
Add difference in movement depending on if player is in air or not

ChristopherKallstrom committed:
Add reduced jump height if not holding space after jumping

ChristopherKallstrom committed:
Add scale to deltaTime depending on window size (works for now, not a good solution in the long run).

ChristopherKallstrom committed:
Remove debugging code

ChristopherKallstrom committed:
Add ability to move player to cursor with right click

ChristopherKallstrom committed:
Add player ability to break vertical velocity by jumping

ChristopherKallstrom committed:
Remove sprinting unintentionally increasing jump height

ChristopherKallstrom committed:
Change high jump to only work when ascending

ChristopherKallstrom committed:
Add unintentionally removed bottom platform

ChristopherKallstrom committed:
Gotta go fast

Smaugmuel commented on GitWorkflow/GitWorkflow/main.cpp:
```diff
@@ -12,13 +12,15 @@ using Time = std::chrono::time_point<Clock>;
 int main()
 {
 	sf::RenderWindow window(sf::VideoMode(WINDOW_WIDTH, WINDOW_HEIGHT), "SFML works!");
+	const float windowScale = (WINDOW_WIDTH / WINDOW_HEIGHT) / (1080 / 720);
```
Division by integers, three times, expecting a float? Hmmmmm.

Smaugmuel commented on GitWorkflow/GitWorkflow/main.cpp:
```diff
@@ -12,13 +12,15 @@ using Time = std::chrono::time_point<Clock>;
 
 	level.create<EndPoint>(sf::Vector2f(20.0f, 20.0f), sf::Vector2f(300.0f, 100.0f));
 	level.create<Player>(sf::Vector2f(20.0f, 20.0f), sf::Vector2f(100.0f, 100.0f));
+	level.create<Platform>(sf::Vector2f(WINDOW_WIDTH, 10.0f), sf::Vector2f(0.0f, WINDOW_HEIGHT));
```
I think we should see the bottom platform, so maybe move it "one platform height" up.

Smaugmuel commented on GitWorkflow/GitWorkflow/main.cpp:
```diff
@@ -33,13 +35,14 @@ int main()
+			if (event.type == sf::Event::MouseButtonReleased)
 			{
-				level.create<Platform>(platformSize, static_cast<sf::Vector2f>(sf::Mouse::getPosition(window) - static_cast<sf::Vector2i>(platformSize / 2.0f)));
+				if(event.mouseButton.button == sf::Mouse::Left) level.create<Platform>(platformSize, static_cast<sf::Vector2f>(sf::Mouse::getPosition(window) - static_cast<sf::Vector2i>(platformSize / 2.0f)));
```
Reduce line length.

Smaugmuel commented on GitWorkflow/GitWorkflow/main.cpp:
```diff
@@ -33,13 +35,14 @@ int main()
 			{
-				level.create<Platform>(platformSize, static_cast<sf::Vector2f>(sf::Mouse::getPosition(window) - static_cast<sf::Vector2i>(platformSize / 2.0f)));
+				if(event.mouseButton.button == sf::Mouse::Left) level.create<Platform>(platformSize, static_cast<sf::Vector2f>(sf::Mouse::getPosition(window) - static_cast<sf::Vector2i>(platformSize / 2.0f)));
+				else if (event.mouseButton.button == sf::Mouse::Right) { level.getPlayer().setPosition(static_cast<sf::Vector2f>(sf::Mouse::getPosition(window))); }
```
Reduce line length.

Smaugmuel commented on GitWorkflow/GitWorkflow/Player.cpp:
```diff
@@ -21,19 +21,28 @@ void Player::collide(const Platform& platform)
 void Player::update(const float dt)
 {
-	m_velocity.x = 0.0f;
+	if (std::roundf(m_velocity.y) == 0.0f && (sf::Keyboard::isKeyPressed(sf::Keyboard::LShift) || sf::Keyboard::isKeyPressed(sf::Keyboard::RShift))) m_speed = m_baseSpeed * 1.5f;
```
Reduce line length.

Smaugmuel commented on GitWorkflow/GitWorkflow/Player.cpp:
```diff
@@ -21,19 +21,28 @@ void Player::collide(const Platform& platform)
 
-	if (sf::Keyboard::isKeyPressed(sf::Keyboard::A)) m_velocity.x -= m_speed;
-	if (sf::Keyboard::isKeyPressed(sf::Keyboard::D)) m_velocity.x += m_speed;
+	if (sf::Keyboard::isKeyPressed(sf::Keyboard::A)) { if (std::abs(m_velocity.x) < m_speed * (std::roundf(m_velocity.y) == 0.0f ? 1.0f : 0.5f)) m_velocity.x -= m_acceleration * dt * (std::roundf(m_velocity.y) == 0.0f ? 1.0f : 0.5f); }
```
Reduce line length.

Smaugmuel commented on GitWorkflow/GitWorkflow/Player.cpp:
```diff
@@ -21,19 +21,28 @@ void Player::collide(const Platform& platform)
+	else if (m_velocity.x < 0.0f) m_velocity.x += m_acceleration * dt;
 
-	if (sf::Keyboard::isKeyPressed(sf::Keyboard::Space) && !m_hasJumped)
+	if (sf::Keyboard::isKeyPressed(sf::Keyboard::D)) { if (std::abs(m_velocity.x) < m_speed * (std::roundf(m_velocity.y) == 0.0f ? 1.0f : 0.5f)) m_velocity.x += m_acceleration * dt * (std::roundf(m_velocity.y) == 0.0f ? 1.0f : 0.5f); }
```
Reduce line length.

Smaugmuel commented on GitWorkflow/GitWorkflow/Player.cpp:
```diff
@@ -21,19 +21,28 @@ void Player::collide(const Platform& platform)
+			m_hasJumped = true;
+		}
 	}
 
```
Remove newline between if and if-else

Smaugmuel commented on GitWorkflow/GitWorkflow/Player.cpp:
```diff
@@ -21,19 +21,28 @@ void Player::collide(const Platform& platform)
 
-	move(sf::Vector2f(0.0f, 9.82f) * (dt * dt / 2.0f));
-	m_velocity.y += 9.82f * dt;
+	else if (m_hasJumped && m_velocity.y < 0.0f) { m_velocity.y += m_speed * dt;  }
```
This looks strange. Only applying this velocity while the player is moving up?

ChristopherKallstrom replied:
It should only apply when the player is ascending ergo when his vertical velocity is negative. Otherwise it will lead to the player falling faster, unless the Space button is being held down.

Smaugmuel commented on GitWorkflow/GitWorkflow/Player.cpp:
```diff
@@ -21,19 +21,28 @@ void Player::collide(const Platform& platform)
+	else if (m_hasJumped && m_velocity.y < 0.0f) { m_velocity.y += m_speed * dt;  }
+
+	move(sf::Vector2f(0.0f, 9.82f) * 24.0f * (dt * dt / 2.0f));
+	m_velocity.y += 9.82f * 24.0f * dt; // * 4 IS FOR TESTING PURPOSES ONLY
```
Update comment

Smaugmuel commented on GitWorkflow/GitWorkflow/Level.cpp:
```diff
@@ -1,10 +1,17 @@
 {
 }
 
+Player& Level::getPlayer() const
+{
+	assert(m_player);
+	return *m_player;
+}
+
```
Personally, I'm not a fan of asserts in product code.
I would prefer std::optionals or pointers in this case.

Smaugmuel reviewed (changes requested):
A very good basis for movement! Values need tweaking and code needs to be cleaned instead of being a jungle of random values.
Try and improve the code quality for movement, it is a non-reusable mess and looks worse.
Also, try and limit the line length. if-statement scopes on the same line as the statement should be very small and only one instruction.

ChristopherKallstrom committed:
Minor code refactoring

Smaugmuel commented on GitWorkflow/GitWorkflow/Level.hpp:
```diff
@@ -22,7 +22,7 @@ class Level final : public sf::Drawable
 	}
 
-	Player& getPlayer() const;
+	Player* getPlayer() const;
```
For const correctness here, add const after asterix.

Smaugmuel commented on GitWorkflow/GitWorkflow/Level.cpp:
```diff
@@ -6,10 +6,9 @@ Level::Level()
 }
 
-Player& Level::getPlayer() const
+Player* Level::getPlayer() const
```
For const correctness here, add const after asterix.

Smaugmuel commented on GitWorkflow/GitWorkflow/main.cpp:
```diff
@@ -12,15 +12,15 @@ using Time = std::chrono::time_point<Clock>;
 	level.create<EndPoint>(sf::Vector2f(20.0f, 20.0f), sf::Vector2f(300.0f, 100.0f));
 	level.create<Player>(sf::Vector2f(20.0f, 20.0f), sf::Vector2f(100.0f, 100.0f));
-	level.create<Platform>(sf::Vector2f(WINDOW_WIDTH, 10.0f), sf::Vector2f(0.0f, WINDOW_HEIGHT));
+	level.create<Platform>(sf::Vector2f(WINDOW_WIDTH, 10.0f), sf::Vector2f(0.0f, WINDOW_HEIGHT - 10.0f));
```
As long as the bottom platform is as high as other platforms, use platformSize.y here instead.

Smaugmuel commented on GitWorkflow/GitWorkflow/Player.cpp:
```diff
@@ -21,29 +21,43 @@ void Player::collide(const Platform& platform)
+	if (m_collidedLastFrame && 
+		(sf::Keyboard::isKeyPressed(sf::Keyboard::LShift) || sf::Keyboard::isKeyPressed(sf::Keyboard::RShift))) 
+		m_speed = m_baseSpeed * m_sprintModifier;
 	else m_speed = m_baseSpeed;
```
I would have created a temporary variable named isSprinting, and set it to true if conditions are met.
Then created a const float named "speed", similar to "control" below, and not stored m_speed in Player. Thoughts?

ChristopherKallstrom replied:
Design was changed in latest branch version, so this shouldn't be an issue anymore.

Smaugmuel commented on GitWorkflow/GitWorkflow/Player.cpp:
```diff
@@ -21,29 +21,43 @@ void Player::collide(const Platform& platform)
+	if (sf::Keyboard::isKeyPressed(sf::Keyboard::A)) 
+	{ 
+		if (std::abs(m_velocity.x) < m_speed * control) 
+			m_velocity.x -= m_acceleration * control * dt;
```
This can be extracted to a separate function and reused for both A and D key.

ChristopherKallstrom replied:
This should be fixed in most recent branch version.

ChristopherKallstrom committed:
Update movement physics and precision

Add basic time step to not miss collisions

Smaugmuel commented on GitWorkflow/GitWorkflow/Player.cpp:
```diff
@@ -21,29 +21,43 @@ void Player::collide(const Platform& platform)
+		if (std::abs(m_velocity.x) < m_speed * control)
+			m_velocity.x += m_acceleration * dt * control;
+	}
+	else if (m_velocity.x > 0.0f) m_velocity.x -= m_acceleration * control * dt;
```
These two else if statements will cause the player to alternate between a positive and negative (albeit small) x-velocity.
After both keys are checked, resolve velocity in one place instead of one, and if it's less than a specified threshold it should be set to zero.

ChristopherKallstrom replied:
This should be fixed in most recent branch version.

Smaugmuel reviewed (changes requested):
Looks good overall, but some improvements can be made.
Const correctness can be improved in some places, and the x-velocity should be set to zero when low enough.

ChristopherKallstrom committed:
Merge branch 'Movement' of github.com:Smaugmuel/GitWorkflow into Movement

ChristopherKallstrom committed:
Fix requested changes.

Update Player hpp and cpp to resolve old merge issues.

Smaugmuel commented on GitWorkflow/GitWorkflow/Player.cpp:
```diff
@@ -62,23 +62,49 @@ void Player::updateMovement(const float dt)
+	m_movingRight = sf::Keyboard::isKeyPressed(sf::Keyboard::D);
+	m_sprinting = sf::Keyboard::isKeyPressed(sf::Keyboard::LShift);
+	m_jumping = sf::Keyboard::isKeyPressed(sf::Keyboard::Space);
+}
```
Very good that input was separated from the actual movement!
I was gonna suggest something similar for the next sprint.
But they should be renamed to signify that a button is pressed, since "m_movingLeft" etc. looks confusing in the update code.
Something like "m_isMoveLeftButtonDown", or better would increase readability.

Smaugmuel commented on GitWorkflow/GitWorkflow/Player.cpp:
```diff
@@ -62,23 +62,49 @@ void Player::updateMovement(const float dt)
 	}
```
Since Player::updateMovement() actually updates velocity, it should be named accordingly.
I was unable to comment on the function signature itself, hopefully this is acceptable.

Smaugmuel commented on GitWorkflow/GitWorkflow/Player.cpp:
```diff
@@ -62,23 +62,49 @@ void Player::updateMovement(const float dt)
+	else if (m_velocity.x < m_speed)
+		accelerate({ m_acceleration, 0.0f }, dt);
+	else if (m_velocity.x > m_speed && !m_sprinting && isColliding())
+		accelerate({ -m_acceleration, 0.0f }, dt);
```
The (!m_movingLeft) and (!m_movingRight) if-statements are nearly identical.
They can be combined into one, which simply deaccelerates the player regardless of direction.

Smaugmuel commented on GitWorkflow/GitWorkflow/Player.cpp:
```diff
@@ -62,23 +62,49 @@ void Player::updateMovement(const float dt)
+	else if (m_velocity.x > m_speed && !m_sprinting && isColliding())
+		accelerate({ -m_acceleration, 0.0f }, dt);
 
-	if (sf::Keyboard::isKeyPressed(sf::Keyboard::Space) && !m_hasJumped)
```
In Player::accelerate(), I believe the two if-statements can be combined into one, but I haven't confirmed it yet.

Smaugmuel reviewed (changes requested):
Overall a very good improvement.
The combinations of if-statements within Player::accelerate() and Player::updateMovement() can wait if you don't have the time, but the names should be changed as mentioned in my other comments.

ChristopherKallstrom committed:
Fix requested changes.
Change input related variable names to be more clear.
Change updateMovement to updateVelocity to be more clear.
Reuse code in accelerate and updateVelocity functions.

ChristopherKallstrom committed:
Merge branch 'master' of github.com:Smaugmuel/GitWorkflow into Movement
